cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -803,6 +805,100 @@ public void suspendResumeSchedule(String namespaceId, String appId, String sched
+   * @param scheduleSpec the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void addSchedule(ApplicationId applicationId, ScheduleSpecification scheduleSpec)
+    throws NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationSpecification appSpec = store.getApplication(applicationId);
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(applicationId);
+    }
+
+    String scheduleName = scheduleSpec.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec != null) {
+      throw new AlreadyExistsException(
+        new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(), scheduleName));
+    }
+
+    // TODO: Figure out if there a way to make the scheduler update and store update transactional
+    ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
+    String programName = scheduleSpec.getProgram().getProgramName();
+    ProgramId programId = new ProgramId(applicationId, programType, programName);
+    scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule(),
+                       scheduleSpec.getProperties());
+    store.addSchedule(programId, scheduleSpec, false);
+  }
+
+  /**
+   * Update the schedule in an application.
+   *
+   * @param applicationId the application containing the schedule
+   * @param scheduleSpecUpdate update to the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void updateSchedule(ApplicationId applicationId, ScheduleSpecification scheduleSpecUpdate)
+    throws NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationSpecification appSpec = store.getApplication(applicationId);
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(applicationId);
+    }
+
+    String scheduleName = scheduleSpecUpdate.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec == null) {
+      throw new NotFoundException(new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(),
+                                                 scheduleName));
+    }
+
+    // TODO: Overlay the schedule spec update on to the existing spec
+    // TODO: Also handle schedule type change
+    ScheduleSpecification finalScheduleSpec = scheduleSpecUpdate;
+    // TODO: Figure out if there a way to make the scheduler update and store update transactional
+    ProgramType programType = ProgramType.valueOfSchedulableType(finalScheduleSpec.getProgram().getProgramType());
```

Comment:
does this mean that an update of a schedule may also change the program (workflow) that is being scheduled?